SPEC.md

```diff
@@ -880,10 +880,8 @@ multi-line-string-body := (('"' | '""')? string-character)*
 string-character := '\\' (["\\bfnrts] | 'u{' hex-unicode '}') | ws-escape | [^\\"] - disallowed-literal-code-points
 ws-escape := '\\' (unicode-space | newline)+
 hex-digit := [0-9a-fA-F]
-hex-unicode := hex-digit{1, 6} - surrogates
-surrogates := [dD][8-9a-fA-F]hex-digit{2}
-// U+D800-DFFF: D  8         00
-//              D  F         FF
+hex-unicode := [\u{0}-\u{10FFFF}] - surrogate  // Unicode Scalar Value₁₆, leading 0s allowed as long as length ≤ 6
```

Comment:
I think `\u` is used to indicate unicode symbols in this metalanguage (see `bom := '\u{FEFF}'` below), so this actually means that hex-unicode is any unicode scalar value, not hexadecimal digits. `[...]` itself can represent only one character.

Reply:
Yes, that **could** be more confusing, just thought that the name + comment would resolve it since this is still not a strict formal grammar?

Otherwise need to complicate the spec more with a few extra regex rules similar to the one removed for >10FFFF and for many 0s, which will hurt readability?

Reply:
Personally I think it would be best to keep `hex-digit{1,6}` here (e.g. same as in https://doc.rust-lang.org/reference/tokens.html#character-literals) as it had been previously and then clearly state how it is allowed to resolve the unicode escape

Reply:
But the lack of clarity is what prompted this! I realized that the syntax is wrong when FFFFFF was highlighted fine because it followed this simple rule of 6. If you think this alt is also not clear, I'd then just add a few more rules, let it be more complicated, but less error-prone

Also, the rust tokens are incomplete, there are rules that are not listed there but the compiler warns you about (think this happens after tokenization), so it's not a great reference for this case

Reply:
Perhaps clarity could be solved by placing a comment there and modifying the spec text? The grammar shouldn't necessarily block all escapes that cannot be resolved, I think, i.e. resolving escapes is not necessarily total.

Reply:
We should be consistent: if we filter out the surrogates in the grammar, then we should also filter out > 10ffff.

IMO either we end up with something like

```
hex-unicode := hex-digit{1, 6} - surrogates - above-max-scalar
surrogates := 0{0,2}[dD][8-9a-fA-F]hex-digit{2}
// U+D800-DFFF: D  8         00
//              D  F         FF
above-max-scalar = '1' [1-9a-fA-F] hex-digit{4} | [2-9a-fA-F] hex-digit{5}
```

or

```
hex-unicode := (hex-digit{1, 5} | '0' hex-digit{5} | '10' hex-digit{4}) - surrogates
surrogates := '0'{0,2} [dD] [8-9a-fA-F] hex-digit{2}
// U+D800-DFFF: D  8         00
//              D  F         FF
```

(in which case we should really define what `{<number>}` and `{<number>,<number>}` mean in the "Grammar language" section at the bottom), or we define `hex-unicode` via a reference to the prose where it is explained that unicode hex values can be left-padded with zeros up to a length of six, that you aren't allowed to encode non-scalar values, &hellip;

Reply:
I agree we should explicitly filter everything out, otherwise it's too easy to miss

What do you think about trying to define it via an explicit range "D800-DFFF" in some format instead of the multiple regexes that amount to the same? Or better regexes + range in the comments?